File: canvas-progress-tracker/obsidian-plugin/src/services/TextbookContextService.ts
/**
 * Textbook Context Service - Canvas Learning System Cross-Canvas Associations
 *
 * Service for retrieving textbook context from associated Canvas files.
 * Implements Story 16.5: Agent引用教材上下文
 *
 * @module TextbookContextService
 * @version 1.0.0
 *
 * ✅ Verified from @graphiti Skill (search, hybrid search pattern)
 * ✅ Verified from Story 16.5 Dev Notes (1s timeout, graceful degradation)
 */

import { App, Notice } from 'obsidian';
import type { GraphitiAssociationService } from './GraphitiAssociationService';
import type { AssociationConfigService } from './AssociationConfigService';

/**
 * Textbook context result
 * ✅ Verified from Story 16.5 Dev Notes (TextbookContext interface)
 */
export interface TextbookContext {
    textbook_canvas: string;
    section_name: string;
    node_id: string;
    relevance_score: number;
    content_preview: string;
    link_format: string;  // Obsidian internal link format
}

/**
 * Prerequisite concept
 */
export interface Prerequisite {
    concept_name: string;
    source_canvas: string;
    node_id: string;
    importance: 'required' | 'recommended' | 'optional';
}

/**
 * Full context result
 */
export interface FullTextbookContext {
    contexts: TextbookContext[];
    prerequisites: Prerequisite[];
    query_time_ms: number;
    timed_out: boolean;
}

/**
 * Service configuration
 */
interface TextbookContextConfig {
    timeout: number;          // Default: 1000ms
    maxResults: number;       // Default: 3
    minRelevanceScore: number; // Default: 0.5
}

/**
 * Default configuration
 * ✅ Verified from Story 16.5 Dev Notes (1s timeout)
 */
const DEFAULT_CONFIG: TextbookContextConfig = {
    timeout: 1000,         // 1 second timeout
    maxResults: 3,
    minRelevanceScore: 0.5
};

/**
 * Textbook Context Service
 *
 * Retrieves relevant textbook content for Agent context injection.
 * Implements graceful degradation with 1-second timeout.
 *
 * ✅ Verified from Story 16.5 Dev Notes (TextbookContextService pattern)
 */
export class TextbookContextService {
    private app: App;
    private graphitiService: GraphitiAssociationService | null;
    private configService: AssociationConfigService;
    private config: TextbookContextConfig;

    /**
     * Creates a new TextbookContextService
     *
     * @param app - Obsidian App instance
     * @param graphitiService - Optional Graphiti service (null for local-only mode)
     * @param configService - Association config service
     * @param config - Optional configuration
     */
    constructor(
        app: App,
        graphitiService: GraphitiAssociationService | null,
        configService: AssociationConfigService,
        config?: Partial<TextbookContextConfig>
    ) {
        this.app = app;
        this.graphitiService = graphitiService;
        this.configService = configService;
        this.config = { ...DEFAULT_CONFIG, ...config };
    }

    /**
     * Get textbook context for a query
     *
     * ✅ Verified from Story 16.5 Dev Notes (timeout protection pattern)
     *
     * @param canvasPath - Current Canvas path
     * @param query - User query
     * @returns Promise<FullTextbookContext | null>
     */
    async getTextbookContext(
        canvasPath: string,
        query: string
    ): Promise<FullTextbookContext | null> {
        const startTime = Date.now();

        try {
            // Use Promise.race for timeout protection
            const result = await Promise.race<FullTextbookContext | null>([
                this.fetchContext(canvasPath, query),
                this.createTimeoutPromise()
            ]);

            if (result === null) {
                // Timeout occurred
                console.warn('[TextbookContextService] Query timed out for:', canvasPath);
                return {
                    contexts: [],
                    prerequisites: [],
                    query_time_ms: Date.now() - startTime,
                    timed_out: true
                };
            }

            result.query_time_ms = Date.now() - startTime;
            return result;

        } catch (error) {
            console.warn('[TextbookContextService] Error fetching context:', error);
            return null;
        }
    }

    /**
     * Fetch context from associated Canvas files
     */
    private async fetchContext(
        canvasPath: string,
        query: string
    ): Promise<FullTextbookContext> {
        const contexts: TextbookContext[] = [];
        const prerequisites: Prerequisite[] = [];

        // 1. Get local associations
        const associations = await this.configService.getAssociations(canvasPath);

        // 2. Filter for REFERENCES type (教材关联)
        const textbookAssociations = associations.filter(
            a => a.association_type === 'references'
        );

        // 3. Search in each textbook Canvas
        for (const assoc of textbookAssociations) {
            const textbookContexts = await this.searchInCanvas(
                assoc.target_canvas,
                query
            );
            contexts.push(...textbookContexts);
        }

        // 4. Get prerequisites
        const prereqAssociations = associations.filter(
            a => a.association_type === 'prerequisite'
        );

        for (const assoc of prereqAssociations) {
            prerequisites.push({
                concept_name: this.getCanvasDisplayName(assoc.target_canvas),
                source_canvas: assoc.target_canvas,
                node_id: assoc.association_id,
                importance: 'required'
            });
        }

        // 5. Try Graphiti for additional context (if available)
        if (this.graphitiService) {
            try {
                const graphitiContexts = await this.fetchFromGraphiti(canvasPath, query);
                contexts.push(...graphitiContexts);
            } catch (error) {
                console.warn('[TextbookContextService] Graphiti search failed:', error);
                // Continue with local results
            }
        }

        // 6. Sort and limit results
        const sortedContexts = contexts
            .sort((a, b) => b.relevance_score - a.relevance_score)
            .slice(0, this.config.maxResults);

        return {
            contexts: sortedContexts,
            prerequisites,
            query_time_ms: 0,
            timed_out: false
        };
    }

    /**
     * Search for relevant content in a Canvas file
     *
     * ✅ Verified from @obsidian-canvas Skill (Canvas JSON parsing)
     */
    private async searchInCanvas(
        canvasPath: string,
        query: string
    ): Promise<TextbookContext[]> {
        const contexts: TextbookContext[] = [];

        try {
            const file = this.app.vault.getAbstractFileByPath(canvasPath);
            if (!file) return [];

            // Read Canvas JSON
            const content = await this.app.vault.read(file as any);
            const canvasData = JSON.parse(content);

            if (!canvasData.nodes || !Array.isArray(canvasData.nodes)) {
                return [];
            }

            // Search nodes for query match
            const queryLower = query.toLowerCase();
            for (const node of canvasData.nodes) {
                if (node.type !== 'text') continue;

                const nodeText = node.text || '';
                const nodeLower = nodeText.toLowerCase();

                // Simple relevance scoring based on query match
                if (nodeLower.includes(queryLower)) {
                    const relevance = this.calculateRelevance(query, nodeText);

                    if (relevance >= this.config.minRelevanceScore) {
                        contexts.push({
                            textbook_canvas: canvasPath,
                            section_name: this.extractSectionName(nodeText),
                            node_id: node.id,
                            relevance_score: relevance,
                            content_preview: nodeText.slice(0, 100),
                            link_format: this.formatInternalLink(canvasPath, node.id, nodeText)
                        });
                    }
                }
            }

        } catch (error) {
            console.warn('[TextbookContextService] Failed to search canvas:', canvasPath, error);
        }

        return contexts;
    }

    /**
     * Fetch additional context from Graphiti
     */
    private async fetchFromGraphiti(
        canvasPath: string,
        query: string
    ): Promise<TextbookContext[]> {
        if (!this.graphitiService) return [];

        try {
            const searchResult = await this.graphitiService.searchCrossCanvasConcepts(query);

            return searchResult.nodes
                .filter(node => node.entity_type === 'LearningNode' || node.entity_type === 'ConceptNode')
                .map(node => ({
                    textbook_canvas: canvasPath, // Would need graph traversal for actual canvas
                    section_name: node.name,
                    node_id: node.uuid,
                    relevance_score: 0.7,
                    content_preview: node.summary || '',
                    link_format: ''
                }))
                .slice(0, 5);

        } catch {
            return [];
        }
    }

    /**
     * Calculate relevance score between query and text
     */
    private calculateRelevance(query: string, text: string): number {
        const queryLower = query.toLowerCase();
        const textLower = text.toLowerCase();

        // Exact match
        if (textLower === queryLower) return 1.0;

        // Contains full query
        if (textLower.includes(queryLower)) {
            // Higher score for shorter text (more focused)
            const lengthFactor = Math.min(1, 100 / text.length);
            return 0.7 + (0.2 * lengthFactor);
        }

        // Word overlap
        const queryWords = queryLower.split(/\s+/);
        const textWords = new Set(textLower.split(/\s+/));

        let matchCount = 0;
        for (const word of queryWords) {
            if (textWords.has(word) || [...textWords].some(w => w.includes(word))) {
                matchCount++;
            }
        }

        return matchCount / queryWords.length;
    }

    /**
     * Extract section name from node text
     */
    private extractSectionName(text: string): string {
        // Try to find a header pattern
        const headerMatch = text.match(/^#+\s*(.+)$/m);
        if (headerMatch) {
            return headerMatch[1].trim();
        }

        // Use first line as section name
        const firstLine = text.split('\n')[0];
        if (firstLine.length > 50) {
            return firstLine.slice(0, 47) + '...';
        }

        return firstLine || 'Unknown Section';
    }

    /**
     * Format Obsidian internal link
     *
     * ✅ Verified from @obsidian-canvas Skill (Internal link syntax)
     */
    private formatInternalLink(canvasPath: string, nodeId: string, text: string): string {
        const displayName = this.extractSectionName(text);
        const canvasName = this.getCanvasDisplayName(canvasPath);

        // Format: [[Canvas#节点ID|显示名]]
        return `[[${canvasPath}#${nodeId}|${canvasName} > ${displayName}]]`;
    }

    /**
     * Get display name from Canvas path
     */
    private getCanvasDisplayName(path: string): string {
        const name = path.split('/').pop() || path;
        return name.replace('.canvas', '');
    }

    /**
     * Create timeout promise
     */
    private createTimeoutPromise(): Promise<null> {
        return new Promise(resolve => {
            setTimeout(() => resolve(null), this.config.timeout);
        });
    }

    /**
     * Build Agent prompt with textbook context
     *
     * ✅ Verified from Story 16.5 Dev Notes (build_agent_prompt pattern)
     *
     * @param basePrompt - Original Agent prompt
     * @param context - Textbook context
     * @returns Enhanced prompt with context
     */
    buildAgentPrompt(
        basePrompt: string,
        context: FullTextbookContext | null
    ): string {
        if (!context || (context.contexts.length === 0 && context.prerequisites.length === 0)) {
            return basePrompt;
        }

        let contextSection = '\n\n## 相关教材参考\n';

        // Add textbook contexts
        for (const ctx of context.contexts) {
            contextSection += `- ${this.getCanvasDisplayName(ctx.textbook_canvas)} > ${ctx.section_name}\n`;
            contextSection += `  内容预览: ${ctx.content_preview}...\n`;
        }

        // Add prerequisites
        if (context.prerequisites.length > 0) {
            contextSection += '\n## 建议先复习\n';
            for (const prereq of context.prerequisites) {
                contextSection += `- ${prereq.concept_name} (${this.getCanvasDisplayName(prereq.source_canvas)})\n`;
            }
        }

        // Add instruction for Agent
        contextSection += `
请在回答中适当引用上述教材内容，格式为"参见教材: {教材名} > {章节名}"。
如果识别到需要前置知识，请提示"建议先复习: {概念名}({来源Canvas})"。
`;

        return basePrompt + contextSection;
    }

    /**
     * Format textbook reference for node metadata
     *
     * @param context - Textbook context
     * @returns Formatted reference string
     */
    formatTextbookRef(context: TextbookContext): string {
        return `参见教材: ${this.getCanvasDisplayName(context.textbook_canvas)} > ${context.section_name}`;
    }

    /**
     * Create clickable links for contexts
     *
     * @param contexts - Array of textbook contexts
     * @returns Markdown formatted links
     */
    createClickableLinks(contexts: TextbookContext[]): string {
        return contexts
            .map(ctx => ctx.link_format)
            .filter(link => link)
            .join('\n');
    }
}

/**
 * Prerequisite Detector
 *
 * Detects and retrieves prerequisite concepts for a given concept.
 * Implements Story 16.5 Task 2.
 *
 * ✅ Verified from Story 16.5 Dev Notes (PrerequisiteDetector pattern)
 */
export class PrerequisiteDetector {
    private app: App;
    private graphitiService: GraphitiAssociationService | null;
    private configService: AssociationConfigService;
    private maxDepth: number = 3;

    /**
     * Creates a new PrerequisiteDetector
     */
    constructor(
        app: App,
        graphitiService: GraphitiAssociationService | null,
        configService: AssociationConfigService
    ) {
        this.app = app;
        this.graphitiService = graphitiService;
        this.configService = configService;
    }

    /**
     * Detect prerequisites for a concept
     *
     * ✅ Verified from Story 16.5 Dev Notes (detectPrerequisites pattern)
     *
     * @param concept - Concept name to find prerequisites for
     * @param canvasPath - Current Canvas path
     * @returns Promise<Prerequisite[]>
     */
    async detectPrerequisites(
        concept: string,
        canvasPath: string
    ): Promise<Prerequisite[]> {
        const prerequisites: Prerequisite[] = [];
        const visited = new Set<string>();

        await this.traversePrerequisites(
            canvasPath,
            prerequisites,
            visited,
            0,
            'required'
        );

        return prerequisites;
    }

    /**
     * Traverse prerequisite chain
     */
    private async traversePrerequisites(
        canvasPath: string,
        prerequisites: Prerequisite[],
        visited: Set<string>,
        depth: number,
        importance: 'required' | 'recommended' | 'optional'
    ): Promise<void> {
        if (depth >= this.maxDepth || visited.has(canvasPath)) {
            return;
        }

        visited.add(canvasPath);

        // Get local prerequisites
        const associations = await this.configService.getAssociations(canvasPath);
        const prereqAssocs = associations.filter(a => a.association_type === 'prerequisite');

        for (const assoc of prereqAssocs) {
            if (visited.has(assoc.target_canvas)) continue;

            prerequisites.push({
                concept_name: this.getCanvasDisplayName(assoc.target_canvas),
                source_canvas: assoc.target_canvas,
                node_id: assoc.association_id,
                importance: this.determineImportance(depth)
            });

            // Recurse for transitive prerequisites
            await this.traversePrerequisites(
                assoc.target_canvas,
                prerequisites,
                visited,
                depth + 1,
                this.determineImportance(depth + 1)
            );
        }

        // Also check Graphiti if available
        if (this.graphitiService && depth === 0) {
            try {
                const graphitiPrereqs = await this.fetchGraphitiPrerequisites(canvasPath);
                for (const prereq of graphitiPrereqs) {
                    if (!visited.has(prereq.source_canvas)) {
                        prerequisites.push(prereq);
                    }
                }
            } catch {
                // Continue without Graphiti
            }
        }
    }

    /**
     * Fetch prerequisites from Graphiti
     */
    private async fetchGraphitiPrerequisites(canvasPath: string): Promise<Prerequisite[]> {
        if (!this.graphitiService) return [];

        // Search for REQUIRES relationships
        const results = await this.graphitiService.searchCrossCanvasConcepts(
            `Prerequisites for ${canvasPath}`
        );

        return results.edges
            .filter(edge => edge.name === 'REQUIRES')
            .map(edge => ({
                concept_name: edge.fact.split(' ')[0] || 'Unknown',
                source_canvas: '', // Would need additional lookup
                node_id: edge.uuid,
                importance: 'recommended' as const
            }));
    }

    /**
     * Determine importance based on depth
     */
    private determineImportance(depth: number): 'required' | 'recommended' | 'optional' {
        if (depth === 0) return 'required';
        if (depth === 1) return 'recommended';
        return 'optional';
    }

    /**
     * Get display name from Canvas path
     */
    private getCanvasDisplayName(path: string): string {
        const name = path.split('/').pop() || path;
        return name.replace('.canvas', '');
    }

    /**
     * Format prerequisites for display
     *
     * @param prerequisites - Array of prerequisites
     * @returns Formatted markdown string
     */
    formatPrerequisites(prerequisites: Prerequisite[]): string {
        if (prerequisites.length === 0) return '';

        let result = '## 建议先复习\n\n';

        // Group by importance
        const required = prerequisites.filter(p => p.importance === 'required');
        const recommended = prerequisites.filter(p => p.importance === 'recommended');
        const optional = prerequisites.filter(p => p.importance === 'optional');

        if (required.length > 0) {
            result += '**必修**:\n';
            for (const prereq of required) {
                result += `- ${prereq.concept_name} (${this.getCanvasDisplayName(prereq.source_canvas)})\n`;
            }
        }

        if (recommended.length > 0) {
            result += '\n**推荐**:\n';
            for (const prereq of recommended) {
                result += `- ${prereq.concept_name} (${this.getCanvasDisplayName(prereq.source_canvas)})\n`;
            }
        }

        if (optional.length > 0) {
            result += '\n**可选**:\n';
            for (const prereq of optional) {
                result += `- ${prereq.concept_name} (${this.getCanvasDisplayName(prereq.source_canvas)})\n`;
            }
        }

        return result;
    }
}

File: backend/app/services/textbook_context_service.py
# Canvas Learning System - Textbook Context Service
# Story FIX-3.1: Python版TextbookContextService
"""
Textbook Context Service for Canvas Learning System.

Provides textbook reference context injection for AI agents.
Reads .canvas-links.json configuration files to find associated
textbook Canvas files and searches for relevant content.

[Source: docs/prd/sprint-change-proposal-20251208.md - Phase 3]
"""

import asyncio
import json
import logging
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any, Dict, List, Optional

logger = logging.getLogger(__name__)


@dataclass
class TextbookContext:
    """
    Textbook context result.

    Attributes:
        textbook_canvas: Path to the textbook Canvas file
        section_name: Name of the matched section
        node_id: ID of the matched node
        relevance_score: Relevance score (0.0-1.0)
        content_preview: Preview of matched content
        page_number: PDF page number (None for non-PDF files)
        file_type: File type - "canvas", "markdown", or "pdf"

    [Source: Story 28.3 - PDF页码链接支持]
    """
    textbook_canvas: str
    section_name: str
    node_id: str
    relevance_score: float
    content_preview: str
    # Story 28.3: PDF page link support
    page_number: Optional[int] = None
    file_type: str = "canvas"  # "canvas" | "markdown" | "pdf"


@dataclass
class Prerequisite:
    """
    Prerequisite concept.

    Attributes:
        concept_name: Name of the prerequisite concept
        source_canvas: Canvas containing the prerequisite
        node_id: Node ID of the prerequisite
        importance: 'required', 'recommended', or 'optional'
    """
    concept_name: str
    source_canvas: str
    node_id: str
    importance: str  # 'required', 'recommended', 'optional'


@dataclass
class FullTextbookContext:
    """
    Full context result including textbooks and prerequisites.

    Attributes:
        contexts: List of textbook contexts (supports PDF page_number and file_type)
        prerequisites: List of prerequisite concepts
        query_time_ms: Time taken for query in milliseconds
        timed_out: Whether the query timed out

    Note:
        Each TextbookContext in contexts may have:
        - page_number: PDF page number for direct page links
        - file_type: "canvas", "markdown", or "pdf"

    [Source: Story 28.3 - PDF页码链接支持]
    """
    contexts: List[TextbookContext] = field(default_factory=list)
    prerequisites: List[Prerequisite] = field(default_factory=list)
    query_time_ms: float = 0.0
    timed_out: bool = False


@dataclass
class TextbookContextConfig:
    """
    Service configuration.

    Attributes:
        timeout: Query timeout in seconds
        max_results: Maximum number of results
        min_relevance_score: Minimum relevance score threshold
    """
    timeout: float = 1.0  # 1 second timeout
    max_results: int = 3
    min_relevance_score: float = 0.5


class TextbookContextService:
    """
    Service for retrieving textbook context from associated Canvas files.

    Implements textbook reference injection for Agent context.
    Reads .canvas-links.json configuration files to find associations.

    [Source: FIX-3.1 Python版TextbookContextService]
    """

    def __init__(
        self,
        canvas_base_path: str,
        config: Optional[TextbookContextConfig] = None
    ):
        """
        Initialize TextbookContextService.

        Args:
            canvas_base_path: Base path to Canvas files (e.g., 笔记库 path)
            config: Optional configuration
        """
        self._base_path = Path(canvas_base_path)
        self._config = config or TextbookContextConfig()
        self._config_cache: Dict[str, Dict[str, Any]] = {}
        logger.debug(f"TextbookContextService initialized with base_path={canvas_base_path}")

    async def get_textbook_context(
        self,
        canvas_path: str,
        query: str
    ) -> Optional[FullTextbookContext]:
        """
        Get textbook context for a query.

        Uses timeout protection for graceful degradation.

        Args:
            canvas_path: Path to the current Canvas file
            query: User query or node content to search for

        Returns:
            FullTextbookContext or None if error occurred
        """
        import time
        start_time = time.time()

        try:
            # Use timeout protection
            result = await asyncio.wait_for(
                self._fetch_context(canvas_path, query),
                timeout=self._config.timeout
            )
            result.query_time_ms = (time.time() - start_time) * 1000
            return result

        except asyncio.TimeoutError:
            logger.warning(f"[TextbookContextService] Query timed out for: {canvas_path}")
            return FullTextbookContext(
                contexts=[],
                prerequisites=[],
                query_time_ms=(time.time() - start_time) * 1000,
                timed_out=True
            )
        except Exception as e:
            logger.warning(f"[TextbookContextService] Error fetching context: {e}")
            return None

    async def _fetch_context(
        self,
        canvas_path: str,
        query: str
    ) -> FullTextbookContext:
        """
        Fetch context from associated Canvas files.

        Args:
            canvas_path: Current Canvas path
            query: Search query

        Returns:
            FullTextbookContext with results
        """
        contexts: List[TextbookContext] = []
        prerequisites: List[Prerequisite] = []

        # 1. Load .canvas-links.json configuration
        associations = await self._get_associations(canvas_path)

        # 2. Filter for 'references' type (教材关联)
        textbook_associations = [
            a for a in associations
            if a.get('association_type') == 'references'
        ]

        # 3. Search in each textbook Canvas
        for assoc in textbook_associations:
            target_canvas = assoc.get('target_canvas', '')
            if target_canvas:
                textbook_contexts = await self._search_in_canvas(
                    target_canvas, query
                )
                contexts.extend(textbook_contexts)

        # 4. Get prerequisites
        prereq_associations = [
            a for a in associations
            if a.get('association_type') == 'prerequisite'
        ]

        for assoc in prereq_associations:
            target_canvas = assoc.get('target_canvas', '')
            prerequisites.append(Prerequisite(
                concept_name=self._get_canvas_display_name(target_canvas),
                source_canvas=target_canvas,
                node_id=assoc.get('association_id', ''),
                importance='required'
            ))

        # 5. Sort and limit results
        sorted_contexts = sorted(
            contexts,
            key=lambda c: c.relevance_score,
            reverse=True
        )[:self._config.max_results]

        return FullTextbookContext(
            contexts=sorted_contexts,
            prerequisites=prerequisites,
            query_time_ms=0.0,
            timed_out=False
        )

    async def _get_associations(self, canvas_path: str) -> List[Dict[str, Any]]:
        """
        Load associations from .canvas-links.json.

        Args:
            canvas_path: Path to Canvas file

        Returns:
            List of association dictionaries
        """
        config = await self._load_config(canvas_path)
        return config.get('associations', [])

    async def _load_config(self, canvas_path: str) -> Dict[str, Any]:
        """
        Load .canvas-links.json configuration.

        Args:
            canvas_path: Path to Canvas file

        Returns:
            Configuration dictionary
        """
        config_path = self._get_config_path(canvas_path)

        # Check cache
        if config_path in self._config_cache:
            return self._config_cache[config_path]

        try:
            full_path = self._base_path / config_path
            if full_path.exists():
                content = full_path.read_text(encoding='utf-8')
                config = json.loads(content)
                self._config_cache[config_path] = config
                return config
        except Exception as e:
            logger.warning(f"[TextbookContextService] Failed to load config: {e}")

        # Return empty config if file doesn't exist
        return {'version': '1.0.0', 'associations': [], 'settings': {}}

    def _get_config_path(self, canvas_path: str) -> str:
        """
        Get the .canvas-links.json path for a Canvas file.

        Args:
            canvas_path: Canvas file path

        Returns:
            Config file path
        """
        # Config file is in the same directory as the Canvas file
        canvas_p = Path(canvas_path)
        if canvas_p.parent == Path('.'):
            return '.canvas-links.json'
        return str(canvas_p.parent / '.canvas-links.json')

    async def _search_in_canvas(
        self,
        canvas_path: str,
        query: str
    ) -> List[TextbookContext]:
        """
        Search for relevant content in a Canvas file.

        Args:
            canvas_path: Path to Canvas file
            query: Search query

        Returns:
            List of TextbookContext results
        """
        contexts: List[TextbookContext] = []

        try:
            full_path = self._base_path / canvas_path
            if not full_path.exists():
                return []

            # Read Canvas JSON
            content = full_path.read_text(encoding='utf-8')
            canvas_data = json.loads(content)

            nodes = canvas_data.get('nodes', [])
            if not isinstance(nodes, list):
                return []

            # Search nodes for query match
            query_lower = query.lower()
            for node in nodes:
                if node.get('type') != 'text':
                    continue

                node_text = node.get('text', '')
                node_lower = node_text.lower()

                # Check if query is in node text
                if query_lower in node_lower:
                    relevance = self._calculate_relevance(query, node_text)

                    if relevance >= self._config.min_relevance_score:
                        contexts.append(TextbookContext(
                            textbook_canvas=canvas_path,
                            section_name=self._extract_section_name(node_text),
                            node_id=node.get('id', ''),
                            relevance_score=relevance,
                            content_preview=node_text[:100]
                        ))

        except Exception as e:
            logger.warning(f"[TextbookContextService] Failed to search canvas {canvas_path}: {e}")

        return contexts

    def _calculate_relevance(self, query: str, text: str) -> float:
        """
        Calculate relevance score between query and text.

        Args:
            query: Search query
            text: Text to match against

        Returns:
            Relevance score (0.0-1.0)
        """
        query_lower = query.lower()
        text_lower = text.lower()

        # Exact match
        if text_lower == query_lower:
            return 1.0

        # Contains full query
        if query_lower in text_lower:
            # Higher score for shorter text (more focused)
            length_factor = min(1.0, 100 / len(text))
            return 0.7 + (0.2 * length_factor)

        # Word overlap
        query_words = query_lower.split()
        text_words = set(text_lower.split())

        match_count = 0
        for word in query_words:
            if word in text_words or any(word in tw for tw in text_words):
                match_count += 1

        if query_words:
            return match_count / len(query_words)

        return 0.0

    def _extract_section_name(self, text: str) -> str:
        """
        Extract section name from node text.

        Args:
            text: Node text

        Returns:
            Section name
        """
        import re

        # Try to find a header pattern
        header_match = re.search(r'^#+\s*(.+)$', text, re.MULTILINE)
        if header_match:
            return header_match.group(1).strip()

        # Use first line as section name
        first_line = text.split('\n')[0] if text else ''
        if len(first_line) > 50:
            return first_line[:47] + '...'

        return first_line or 'Unknown Section'

    def _get_canvas_display_name(self, path: str) -> str:
        """
        Get display name from Canvas path.

        Args:
            path: Canvas file path

        Returns:
            Display name
        """
        name = Path(path).name
        return name.replace('.canvas', '')

    def build_agent_prompt(
        self,
        base_prompt: str,
        context: Optional[FullTextbookContext]
    ) -> str:
        """
        Build Agent prompt with textbook context.

        Args:
            base_prompt: Original Agent prompt
            context: Textbook context

        Returns:
            Enhanced prompt with context
        """
        if not context or (len(context.contexts) == 0 and len(context.prerequisites) == 0):
            return base_prompt

        context_section = '\n\n## 相关教材参考\n'

        # Add textbook contexts
        for ctx in context.contexts:
            display_name = self._get_canvas_display_name(ctx.textbook_canvas)
            context_section += f'- {display_name} > {ctx.section_name}\n'
            context_section += f'  内容预览: {ctx.content_preview}...\n'

        # Add prerequisites
        if context.prerequisites:
            context_section += '\n## 建议先复习\n'
            for prereq in context.prerequisites:
                display_name = self._get_canvas_display_name(prereq.source_canvas)
                context_section += f'- {prereq.concept_name} ({display_name})\n'

        # Add instruction for Agent
        context_section += '''
请在回答中适当引用上述教材内容，格式为"参见教材: {教材名} > {章节名}"。
如果识别到需要前置知识，请提示"建议先复习: {概念名}({来源Canvas})"。
'''

        return base_prompt + context_section

    def format_textbook_ref(self, context: TextbookContext) -> str:
        """
        Format textbook reference for node metadata.

        Args:
            context: Textbook context

        Returns:
            Formatted reference string
        """
        display_name = self._get_canvas_display_name(context.textbook_canvas)
        return f'参见教材: {display_name} > {context.section_name}'

    def invalidate_cache(self, canvas_path: Optional[str] = None) -> None:
        """
        Invalidate configuration cache.

        Args:
            canvas_path: Specific Canvas path to invalidate, or None for all
        """
        if canvas_path:
            config_path = self._get_config_path(canvas_path)
            self._config_cache.pop(config_path, None)
        else:
            self._config_cache.clear()

    async def sync_mounted_textbook(
        self,
        canvas_path: str,
        association: Dict[str, Any]
    ) -> str:
        """
        Sync a mounted textbook to .canvas-links.json.

        Creates or updates the configuration file with the new association.
        This bridges the frontend localStorage to backend filesystem.

        Args:
            canvas_path: Path to the Canvas file
            association: Association dictionary with textbook info

        Returns:
            Path to the .canvas-links.json file

        [Source: 方案A - 前端同步到后端]
        """
        config_path = self._get_config_path(canvas_path)
        full_path = self._base_path / config_path

        # Load existing config or create new
        try:
            if full_path.exists():
                content = full_path.read_text(encoding='utf-8')
                config = json.loads(content)
            else:
                config = {
                    'version': '1.0.0',
                    'associations': [],
                    'settings': {'auto_link': True}
                }
        except Exception as e:
            logger.warning(f"[TextbookContextService] Failed to load config, creating new: {e}")
            config = {
                'version': '1.0.0',
                'associations': [],
                'settings': {'auto_link': True}
            }

        # Check if association already exists (by ID)
        associations = config.get('associations', [])
        existing_idx = None
        for i, a in enumerate(associations):
            if a.get('association_id') == association.get('association_id'):
                existing_idx = i
                break

        if existing_idx is not None:
            # Update existing
            associations[existing_idx] = association
        else:
            # Add new
            associations.append(association)

        config['associations'] = associations

        # Ensure directory exists
        full_path.parent.mkdir(parents=True, exist_ok=True)

        # Write config file
        full_path.write_text(
            json.dumps(config, ensure_ascii=False, indent=2),
            encoding='utf-8'
        )

        # Invalidate cache
        self._config_cache.pop(config_path, None)

        logger.info(f"[TextbookContextService] Synced textbook to {config_path}")

        return config_path

    async def remove_mounted_textbook(
        self,
        canvas_path: str,
        textbook_id: str
    ) -> bool:
        """
        Remove a mounted textbook from .canvas-links.json.

        Args:
            canvas_path: Path to the Canvas file
            textbook_id: Textbook ID to remove

        Returns:
            True if removed, False if not found

        [Source: 方案A - 前端同步到后端]
        """
        config_path = self._get_config_path(canvas_path)
        full_path = self._base_path / config_path

        if not full_path.exists():
            return False

        try:
            content = full_path.read_text(encoding='utf-8')
            config = json.loads(content)
        except Exception as e:
            logger.warning(f"[TextbookContextService] Failed to load config for removal: {e}")
            return False

        # Find and remove the association
        associations = config.get('associations', [])
        original_count = len(associations)

        # Remove by matching association_id that contains the textbook_id
        associations = [
            a for a in associations
            if not a.get('association_id', '').endswith(textbook_id)
        ]

        if len(associations) == original_count:
            return False  # Nothing removed

        config['associations'] = associations

        # Write back
        full_path.write_text(
            json.dumps(config, ensure_ascii=False, indent=2),
            encoding='utf-8'
        )

        # Invalidate cache
        self._config_cache.pop(config_path, None)

        logger.info(f"[TextbookContextService] Removed textbook {textbook_id} from {config_path}")

        return True


# Singleton instance
_textbook_context_service: Optional[TextbookContextService] = None


def get_textbook_context_service(
    canvas_base_path: Optional[str] = None
) -> TextbookContextService:
    """
    Get or create the TextbookContextService singleton.

    Args:
        canvas_base_path: Base path to Canvas files

    Returns:
        TextbookContextService instance
    """
    global _textbook_context_service

    if _textbook_context_service is None:
        from app.config import settings
        base_path = canvas_base_path or settings.CANVAS_BASE_PATH
        _textbook_context_service = TextbookContextService(base_path)

    return _textbook_context_service


def reset_textbook_context_service() -> None:
    """Reset the TextbookContextService singleton (for testing)."""
    global _textbook_context_service
    _textbook_context_service = None
